packages/pilet-webpack-config-enchancer/src/index.ts

```diff
@@ -0,0 +1,104 @@
+import { BannerPlugin, DefinePlugin, WebpackPluginInstance } from 'webpack';
+import { setEnvironment, getDefineVariables, getVariables } from './helpers';
+
+export interface PiletWebpackConfigEnchancerOptions {
+  /**
+   * The name of the pilet.
+   */
+  name: string;
+  /**
+   * The version of the pilet.
+   */
+  version: string;
+  /**
+   * The name of the Piral instance / app shell.
+   */
+  piral: string;
+  /**
+   * The schema version. By default, v1 is used.
+   */
+  schema?: 'v0' | 'v1' | 'none';
+  /**
+   * The shared dependencies. By default, these are read from the
+   * Piral instance.
+   */
+  externals?: Array<string>;
+  /**
+   * Additional environment variables to define.
+   */
+  variables?: Record<string, string>;
+}
+
+function getExternals(piral: string) {
+  const shellPkg = require(`${piral}/package.json`);
+  const piralExternals = shellPkg.pilets?.externals ?? [];
+  return [
+    ...piralExternals,
+    '@dbeining/react-atom',
+    '@libre/atom',
+    'history',
+    'react',
+    'react-dom',
+    'react-router',
+    'react-router-dom',
+    'tslib',
+    'path-to-regexp',
+  ];
+}
+
+export const piletWebpackConfigEnchancer = (options: PiletWebpackConfigEnchancerOptions) => compilerOptions => {
+  const environment = process.env.NODE_ENV || 'development';
+  const { name, version, piral, externals = getExternals(piral), schema } = options;
+  const shortName = name.replace(/\W/gi, '');
+  const jsonpFunction = `pr_${shortName}`;
+  const variables = {
+    ...getVariables(name, version, environment),
+    ...options.variables,
+  };
+  const plugins: WebpackPluginInstance[] = [new DefinePlugin(getDefineVariables(variables))];
+
+  if (schema !== 'none') {
+    const bannerSuffix = schema ? `1(${jsonpFunction})` : `0`;
+
+    plugins.push(
+      new BannerPlugin({
+        banner: `//@pilet v:${bannerSuffix}`,
+        entryOnly: true,
+        include: /\.js$/,
+        raw: true,
+      }),
+    );
+  }
+
+  // FIXME: this is a really bad idea, making finding the source of configuration nearly impossible
+  setEnvironment(variables);
```

Comment:
Do we need this for webpack5 ? The `process.env` is not used anyway. The only thing I can think of is the `NODE_ENV`, but here, too, its more about the `DefinePlugin` IIRC.